Monaco editor

MichaelPetrinolis opened the pull request:
Add monaco-editor (local and cdn) to resources, and use it as an editor for text and html fields. It depends on #8265 to load monaco loader and overcome the issue that only one anonymous define call per script file can be defined.

Related to #2345

hishamco commented:
@agriffard finally seems someone take the initiative to introduce Monaco Editor

Thanks MichaelPetrinolis

MichaelPetrinolis commented:
@hishamco thanx, is nice to have monaco in the CMS, especially in the resources, eg. we could use the diffEditor in the audit trail module to display differences and many other use cases. Don't know what to make a gif about, it is just another editor for TextField and HtmlField, with settings to define a json configuration for creating the editor.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Views/HtmlField-Monaco.Edit.cshtml:
```diff
@@ -0,0 +1,54 @@
+
+@{
+    var settings = Model.PartFieldDefinition.GetSettings<HtmlFieldSettings>();
+    var monacoSettings = Model.PartFieldDefinition.GetSettings<TextFieldMonacoEditorSettings>();
```
```suggestion
    var monacoSettings = Model.PartFieldDefinition.GetSettings<HtmlFieldMonacoEditorSettings>();
```

Skrypt commented:
Fun

![image](https://user-images.githubusercontent.com/3228637/104067303-e7dd0080-51d0-11eb-9a99-a0d932553802.png)

Skrypt commented:
![image](https://user-images.githubusercontent.com/3228637/104067560-5cb03a80-51d1-11eb-9e42-180dde7652c8.png)

Mobile seems to work but maybe needs some adjustments.

hishamco commented:
@MichaelPetrinolis or @Skrypt this is really need a demo ;)

hishamco commented:
@Skrypt did you try it in a light theme?

Skrypt commented:
@hishamco Yeah, there are some small CSS glitches here and there but it mostly works.

![image](https://user-images.githubusercontent.com/3228637/104079384-abb99800-51f0-11eb-9525-4fb6e43184e3.png)

scleaver commented:
This is great!

agriffard commented:
@MichaelPetrinolis Could you make a demo during next meeting?

MichaelPetrinolis commented:
@agriffard I'll try to attend the meeting :)

hishamco commented:
> @agriffard I'll try to attend the meeting :)

That's great!! or let someone else demo it in case you can't .. I can't wait for this one ;)

hishamco commented:
@MichaelPetrinolis please fix the conflict

MichaelPetrinolis commented:
@hishamco @agriffard I am afraid that due to personal reasons, I won't be able to attend. If someone else could demo, please do so. Sorry for the inconvenience.

agriffard commented:
- Please add some documentation (at least in ContentFields): Config, settings, shortcuts, ...
- The Editor settings are displayed even if the Monaco editor mode is not selected.
- UI when you search:
![image](https://user-images.githubusercontent.com/703248/105832379-2a238180-5fc8-11eb-878b-b1144c06fdcd.png)
- When I change the setttings `"language": "html"` to `"language": "javascript"` and I reedit, it is displaying `"language": "html"` again

sebastienros commented:
Great demo about it today, thanks @MichaelPetrinolis

agriffard commented:
Do you plan to also add it to HtmlBodyPart?

MichaelPetrinolis commented:
@sebastienros sorry for not attending... FYI, I tested the json schema and works. Used a tool to extract Monaco configuration schema from source code and use it in the Monaco settings editor. The only issue I encountered was loading editor from CDN, CORS did not allow loading the above schema.
@agriffard I will fix the mentioned issues and add it to HtmlBodyPart. 

Thanx.

agriffard commented:
Please solve the conflicts.

agriffard commented:
Build error:
`Error: D:\a\OrchardCore\OrchardCore\test\OrchardCore.Tests\ResourceManagement\ResourceManagerTests.cs(334,21): error CS0111: Type 'ResourceManagerTests' already defines a member called 'RequireCircularDependenciesShouldThrowException' with the same parameter types [D:\a\OrchardCore\OrchardCore\test\OrchardCore.Tests\OrchardCore.Tests.csproj]`

MichaelPetrinolis commented:
I added typescript-json-schema as a devDependency and use it to create the IStandaloneEditorConstructionOptions.json schema
```
.\node_modules\.bin\typescript-json-schema 'node_modules/monaco-editor/monaco.d.ts' 'monaco.editor.IStandaloneEditorConstructionOptions' -o wwwroot/Scripts/monaco/IStandaloneEditorConstructionOptions.json
```
this is how it looks when editing settings
![image](https://user-images.githubusercontent.com/9623255/108637805-f3406e80-7494-11eb-8a83-973deb4a643c.png)

I also fixed the css issue (there is still an animation with border-color when clicking to expand the replace section)

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Resources/ResourceManifest.cs:
```diff
@@ -392,6 +402,35 @@ public void BuildManifests(IResourceManifestBuilder builder)
+            //    .SetDependencies("monaco")
+            //    .SetVersion(monacoEditorVersion);
+
+            //manifest
```
Leftover?

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Resources/OrchardCore.Resources.csproj:
```diff
@@ -10,10 +10,32 @@
+    <None Include="wwwroot\Scripts\monaco\cdn.js">
+      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
+    </None>
+    <None Include="wwwroot\Scripts\monaco\dev.js">
```
This confuses me? Do you need to do this?

deanmarcussen reviewed (commented):
Got a merge conflict to resolve still.

And a couple of questions, but think now we've got the resource position sorted out, then this is pretty much there.

MichaelPetrinolis commented:
@deanmarcussen thanx for checking, fixed merge issues and did some minor changes. VS2019 modifies .csproj when you copy/move items through the solution explorer, that's why the items in the project file :)

There is also an issue when using CDN, it does not load the IStandaloneEditorConstructionOptions json schema for intelisense

deanmarcussen commented:
ok great @MichaelPetrinolis 

re this `IStandaloneEditorConstructionOptions` issue, I'm not sure what it does, or why there is a problem with it?

Maybe some options to get this pr moving forward and merged
- Remove the feature that I don't know what is from this pr, to be implemented later
- Fix it?
- If it's crucial, and isn't going to work form a cdn, do we just remove the cdn urls, and always run locally (not a killer, it is for the admin after all, rather than the front end)

is it critical, i.e. can we drop it, and figure out how to implement later, or do we have to have it?

agriffard commented:
As dean said, tell us if you can finish the PR without the `IStandaloneEditorConstructionOptions` if it is a blocker so it can be merged.

Please fix the conflict.

If you add some docs, you can make a link to this demo: https://www.youtube.com/watch?v=lnjdRildsL8

MichaelPetrinolis commented:
@agriffard sure, i will proceed as soon as i get some time. The IStandaloneEditorConstructionOptions is a json schema that is used for providing intellisense when you edit the Monaco initialization options in Admin. The issue is when Monaco is served by CDN, it cannot load the above json schema due to CORS permissions. So intellisense does not work when using CDN resources.

agriffard commented:
Please fix the conflicts (ex: Resources moved to ResourceManagementOptionsConfiguration).

agriffard commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/HtmlFieldMonacoEditorSettingsDriver.cs:
```diff
@@ -0,0 +1,64 @@
+            })
+            .Location("Editor");
+        }
+
```
Remove extra line.

agriffard commented on src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/TextFieldMonacoEditorSettingsDriver.cs:
```diff
@@ -0,0 +1,61 @@
+            })
+            .Location("Editor");
+        }
+
```
Remove extra line.

agriffard commented:
Can you please add a new eol for each file that lacks of it?

agriffard commented on src/OrchardCore.Modules/OrchardCore.Resources/package.json:
```diff
@@ -12,11 +12,15 @@
     "jquery-resizable-dom": "0.35.0",
     "jquery.easing": "1.4.1",
     "js-cookie": "2.2.1",
+    "monaco-editor": "^0.22.3",
```
Please update to https://www.npmjs.com/package/monaco-editor/v/0.23.0 and version in docs/resources/libraries/Readme.md

agriffard commented:
The shortcode modal does not seem to have a way to appear in the Part (no button as in other editors).

If you manage to make it work, please add it for the field editor too.

/cc @deanmarcussen

agriffard commented:
Thank you but, please, choose another shortcut than `Ctrl+c` so we can still copy some text.

MichaelPetrinolis commented:
@agriffard haha, how did this happen! What about alt + c? Any suggestions are welcome

agriffard commented:
> @agriffard haha, how did this happen! What about alt + c? Any suggestions are welcome

`Alt+c` already opens a 'Match case' dialog.

agriffard commented:
Can you please merge dev to fix the conflicts?

agriffard reviewed (approved):
LGTM

deanmarcussen reviewed (approved):
Huge pr. Thanks @MichaelPetrinolis

jptissot commented:
Thanks for this ! Kudos. 

I have a question: Do we need all these languages ?

sebastienros commented:
We should only have the minified versions of the files. I don't think it makes sense to redistribute all sources. We can depend on npm but only copy the min files.

Skrypt commented:
I was trying to find Liquid in this list of languages but it's not supported it requires an extension in vs code so it's the same for the editor here.
